src/python/pants/backend/codegen/tasks/protobuf_gen.py

```diff
@@ -103,55 +86,44 @@ def resolve_deps(self, deps_list, key):
                                   .format(message=e, key=key))
     return deps
 
+  def invalidate_for_files(self):
+    return [self.protobuf_binary]
+
   @property
   def javadeps(self):
     return self.resolve_deps(self.get_options().javadeps, 'javadeps')
 
   @property
-  def pythondeps(self):
-    return self.resolve_deps(self.get_options().pythondeps, 'pythondeps')
-
-  def invalidate_for_files(self):
-    return [self.protobuf_binary]
+  def synthetic_target_type(self):
+    return JavaLibrary
+
+  def synthetic_target_extra_dependencies(self, target):
+    # We need to add in the proto imports jars.
+    jars_address = SyntheticAddress(
+        os.path.relpath(self.codegen_workdir(target), get_buildroot()),
+        target.id + '-rjars')
+    jars_target = self.context.add_new_target(jars_address,
+                                              JarLibrary,
+                                              jars=target.imported_jars,
+                                              derived_from=target)
+    deps = OrderedSet([jars_target])
+    deps.update(self.javadeps)
+    return deps
 
   def is_gentarget(self, target):
     return isinstance(target, JavaProtobufLibrary)
 
-  def is_forced(self, lang):
-    return lang in self.gen_langs
-
-  def genlangs(self):
-    return Target.LANG_DISCRIMINATORS
-
-  def _jars_to_directories(self, target):
-    """Extracts and maps jars to directories containing their contents.
-
-    :returns: a set of filepaths to directories containing the contents of jar.
-    """
-    files = set()
-    jarmap = self.context.products.get('ivy_imports')
-    for folder, names in jarmap.by_target[target].items():
-      for name in names:
-        files.add(self._extract_jar(os.path.join(folder, name)))
-    return files
+  def sources_generated_by_target(self, target):
+    genfiles = []
+    for source in target.sources_relative_to_source_root():
+      path = os.path.join(target.target_base, source)
+      genfiles.extend(self.calculate_genfiles(path, source))
+    return genfiles
 
-  def _extract_jar(self, jar_path):
-    """Extracts the jar to a subfolder of workdir/extracted and returns the path to it."""
-    with open(jar_path, 'rb') as f:
-      outdir = os.path.join(self.workdir, 'extracted', sha1(f.read()).hexdigest())
-    if not os.path.exists(outdir):
-      ZIP.extract(jar_path, outdir)
-      self.context.log.debug('Extracting jar at {jar_path}.'.format(jar_path=jar_path))
-    else:
-      self.context.log.debug('Jar already extracted at {jar_path}.'.format(jar_path=jar_path))
-    return outdir
+  def execute_codegen(self, targets):
+    if not targets:
```

Comment:
Is this an important change?  For some reason I thought that sometimes this task might get called with no targets.  In any case, do we really want to error out? Maybe just print a log message and return.  It isn't anything wrong with this task (maybe the execution engine.)